File: vscode-extension/src/test/services/RuleService.test.ts
import * as assert from 'assert';
import { FolderRuleBuilder, SingleFileRuleBuilder } from '../../services/RuleService';

describe('RuleBuilders', () => {
    
    describe('FolderRuleBuilder', () => {
        it('should maintain file structure and ensure markdown extension', () => {
            const builder = new FolderRuleBuilder('.trae/rules');
            const sourceFiles = [
                { path: 'some/path/rule1.md', content: 'content1' },
                { path: 'rule2', content: 'content2' } // Missing extension
            ];

            const rules = builder.buildRules(sourceFiles, 'Cursor');

            assert.strictEqual(rules.length, 2);
            
            assert.strictEqual(rules[0].path, '.trae/rules/rule1.md'); // path.join uses os separator, assume / for now or normalized
            assert.strictEqual(rules[0].content, 'content1');
            
            assert.strictEqual(rules[1].path, '.trae/rules/rule2.md');
            assert.strictEqual(rules[1].content, 'content2');
        });
    });

    describe('SingleFileRuleBuilder', () => {
        it('should concatenate files into one', () => {
            const builder = new SingleFileRuleBuilder('.codiumai.toml');
            const sourceFiles = [
                { path: 'rule1.md', content: 'content1' },
                { path: 'rule2.md', content: 'content2' }
            ];

            const rules = builder.buildRules(sourceFiles, 'Cursor');

            assert.strictEqual(rules.length, 1);
            const rule = rules[0];
            assert.strictEqual(rule.path, '.codiumai.toml');
            
            // Check content format
            const expectedContent = "content1\n\ncontent2\n\n";
            assert.strictEqual(rule.content, expectedContent);
        });
    });
});

File: jetbrains-plugin/src/test/kotlin/com/oneide/services/RuleServiceTest.kt
package com.oneide.services

import org.junit.Assert.assertEquals
import org.junit.Test

class RuleServiceTest {

    @Test
    fun `test FolderRuleBuilder maintains file structure but ensures markdown extension`() {
        val builder = FolderRuleBuilder(".trae/rules")
        val sourceFiles = listOf(
            RuleFile("some/path/rule1.md", "content1"),
            RuleFile("rule2", "content2") // Missing extension
        )

        val rules = builder.buildRules(sourceFiles, "Cursor")

        assertEquals(2, rules.size)
        assertEquals(".trae/rules/rule1.md", rules[0].path)
        assertEquals("content1", rules[0].content)
        
        assertEquals(".trae/rules/rule2.md", rules[1].path)
        assertEquals("content2", rules[1].content)
    }

    @Test
    fun `test SingleFileRuleBuilder concatenates files`() {
        val builder = SingleFileRuleBuilder(".codiumai.toml")
        val sourceFiles = listOf(
            RuleFile("rule1.md", "content1", 100),
            RuleFile("rule2.md", "content2", 200)
        )

        val rules = builder.buildRules(sourceFiles, "Cursor")

        assertEquals(1, rules.size)
        val rule = rules[0]
        assertEquals(".codiumai.toml", rule.path)
        assertEquals(200, rule.lastModified)
        
        // Check content format
        val expectedContent = "content1\n\ncontent2\n\n"
        assertEquals(expectedContent, rule.content)
    }
}
